### PR TITLE
New onCreate hook for Android in dmSDK (and breaking change for RegisterAndroidOnActivityResultListener)

### PR DESCRIPTION
**Pull request by AGulev**
```
static void OnCreate(JNIEnv *_env, ANativeActivity* activity)
{
  dmAndroid::ThreadAttacher threadAttacher;
  JNIEnv* env = threadAttacher.GetEnv();
  jclass cls = dmAndroid::LoadClass(env, JAVA_CLASS);
  jmethodID method = env->GetStaticMethodID(cls, JAVA_METHOD, "(Landroid/app/Activity;)V");
  env->CallStaticVoidMethod(cls, method, threadAttacher.GetActivity()->clazz, threadAttacher.GetActivity()->clazz);
  dmAndroid::UnregisterAndroidOnCreateListener(OnCreate);
}

struct GlobalRegHooks
{
  GlobalRegHooks() {
    dmAndroid::RegisterAndroidOnCreateListener(OnCreate);
  }
} g_GlobalRegHooks; 

```

More info about a breaking change in `RegisterAndroidOnActivityResultListener` is available here:
https://github.com/defold/extension-gpgs/pull/41/files

### CODE REVIEW

File: engine/glfw/lib/android/android_window.c [hunk 1]
```diff
@@ -704,3 +704,42 @@ Java_com_dynamo_android_DefoldActivity_nativeOnActivityResult(
...
+    }
+    LOGW("onCreate listener not found");
+}
+
+JNIEXPORT void
+Java_com_dynamo_android_DefoldActivity_nativeOnCreate(
+    JNIEnv *env, jobject thiz, jobject activity) {
+
+    int i;
+    for (i = 0; i < g_onCreate_ListenersCount; ++i)
```
Comment (JCash) [inline engine/glfw/lib/android/android_window.c:741]: We usually put the loop variable in the for loop, unless it's actually needed later outside of the loop.

**Review by JCash — APPROVED** [review]

File: engine/extension/src/dmsdk/extension/extension.h [hunk 2]
```diff
@@ -247,6 +247,38 @@ namespace dmExtension
...
+     *
+     * onCreate callback function type.
+     * Used with RegisterAndroidOnCreateListener() and UnregisterAndroidOnCreateListener()
+     *
+     * @typedef
+     * @name OnCreateListener
+     * @param env [type:void*]
+     * @param activity [type:void*]
+     */
+    typedef void (*OnCreateListener)(void* env, void* activity);
```
Comment (JCash) [inline engine/extension/src/dmsdk/extension/extension.h:260]: On second thought, I think the name is a bit too generic.
Perhaps `OnActivityCreate(JNIEnv*, ANativeActivity*)` ?

**Review by JCash — CHANGES_REQUESTED** [review]

File: engine/extension/src/dmsdk/extension/extension.h (hunk 2 above)
Comment (JCash) [reply engine/extension/src/dmsdk/extension/extension.h:260]: To make it build for other platforms:
```
#if defined(ANDROID)
struct JNIEnv;
struct ANativeActivity;
#endif
```

os, like in android.h
```
#if defined(ANDROID)
#include <jni.h>
#include <android_native_app_glue.h> // For ANativeActivity
#endif
```

Comment (JCash) [reply engine/extension/src/dmsdk/extension/extension.h:260]: Now that I mentioned it, I wonder if this function is actually better suited for `android.h`. It's a new file, and we probably want/need similar files for other platforms later on. It allows us to make more specialized functions, without.

Are we really going to need this generic callback on other platforms? If so, we can take it as it comes.
But for now, I'd rather move it into the Android space.

E.g. also rename functions `glfwAndroid*`

What do you think?

Comment (AGulev) [reply engine/extension/src/dmsdk/extension/extension.h:260]: @JCash is it fine to have 
```
#if defined(ANDROID)
#include <jni.h>
#include <android_native_app_glue.h> // For ANativeActivity
#endif
```

in `glfw.h ` ?

File: engine/glfw/lib/android/android_window.c (hunk 1 above)
Comment (AGulev) [reply engine/glfw/lib/android/android_window.c:741]: Thx, I'll fix that

File: engine/glfw/include/GL/glfw.h [hunk 3]
```diff
@@ -578,16 +578,21 @@ GLFWAPI int32_t glfwAndroidWindowOpened();
...
+// Activity control
+typedef void (*glfwactivityresultfun)(void *env, void* activity, int request_code, int result_code, void* result);
+GLFWAPI void glfwAndroidRegisterOnActivityResultListener(glfwactivityresultfun fun);
+GLFWAPI void glfwAndroidUnregisterOnActivityResultListener(glfwactivityresultfun fun);
+
+// onCreate listeners
+typedef void (*glfwoncreatefun)(void *env, void* activity);
+GLFWAPI void glfwAndroidRegisterOnCreateListener(glfwoncreatefun fun);
+GLFWAPI void glfwAndroidUnregisterOnCreateListener(glfwoncreatefun fun);
 #endif
```
Comment (AGulev) [inline engine/glfw/include/GL/glfw.h:591]: //#if defined(ANDROID)

File: engine/extension/src/dmsdk/extension/extension.h (hunk 2 above)
Comment (AGulev) [reply engine/extension/src/dmsdk/extension/extension.h]: I think it's done!
Thank you!!!

**Review by JCash — APPROVED** [review]

**Pull request merged**